extensions/wikia/Flags/models/Flag.class.php

```diff
@@ -0,0 +1,186 @@
+<?php
+
+namespace Flags\Models;
+
+class Flag extends FlagsModel {
+	const FLAG_ACTION_ADD = 'add';
+	const FLAG_ACTION_REMOVE = 'remove';
+
+	private
+		$paramsVerified = false,
+		$status;
+
+	public function getFlagsForPage( $wikiId, $pageId ) {
+		/**
+		 * 1. Get flags data for the page
+		 */
+		$flagsForPage = $this->getFlagsTypesForPage( $wikiId, $pageId );
+
+
+		/**
+		 * 2. Get parameters for flags that need them
+		 */
+		$flagsWithParams = [];
+		foreach ( $flagsForPage as $flag ) {
+			if ( !empty( $flag['flag_params_names'] ) ) {
+				$flagsWithParams[] = $flag['flag_id'];
+			}
+		}
+
+		if ( !empty( $flagsWithParams ) ) {
+			$flagsParams = $this->getParamsForFlags( $flagsWithParams );
+			foreach ( $flagsParams as $flagId => $flagParams ) {
+				$flagsForPage[ $flagId ]['params'] = $flagParams;
+			}
+		}
+
+		var_dump( $flagsForPage );
+
+		/**
+		 * 3. Return the ready table
+		 */
+		return $flagsForPage;
+	}
+
+	private function getFlagsTypesForPage( $wikiId, $pageId ) {
+		$db = $this->getDatabaseForRead();
+
+		$flagsTypes = ( new \WikiaSQL() )
+			->SELECT_ALL( 'flags_to_pages', 'flags_types' )
+			->FROM( 'flags_to_pages' )
+			->INNER_JOIN( 'flags_types' )
+				->ON( 'flags_types.flag_type_id', 'flags_to_pages.flag_type_id' )
+			->WHERE( 'flags_to_pages.wiki_id' )->EQUAL_TO( $wikiId )
+			->AND_( 'flags_to_pages.page_id' )->EQUAL_TO( $pageId )
+			->runLoop( $db, function( &$flagsTypes, $row ) {
+				$flagsTypes[ $row->flag_id ] = get_object_vars( $row );
+			} );
+
+		return $flagsTypes;
+	}
+
+	private function getParamsForFlags( $flagsWithParams ) {
+		$db = $this->getDatabaseForRead();
+
+		$flagsParams = ( new \WikiaSQL() )
+			->SELECT( 'flag_id', 'param_name', 'param_value' )
+			->FROM( 'flags_params' )
+			->WHERE( 'flag_id' )->IN( $flagsWithParams )
+			->runLoop( $db, function( &$flagsParams, $row ) {
+				$flagsParams[ $row->flag_id ][ $row->param_name ] = $row->param_value;
+			} );
+
+		return $flagsParams;
+	}
+
+	public function performAction( $action, $params ) {
+		/**
+		 * The data should be verified in the
+		 * verifyParamsForAction() method before
+		 * performing an action
+		 */
+		if ( !$this->paramsVerified ) {
+			return false;
+		}
+
+		switch ( $action ) {
```

Comment:
Not sure we want to use a switch here, using different methods and possibly building them based on the value of the action variable may be an alternative solution.